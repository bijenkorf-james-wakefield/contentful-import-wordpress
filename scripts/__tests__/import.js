const CONTENTFUL = require("../config");

jest.mock("contentful-management");
jest.mock("fs");

require("dotenv").config({
  path: `.env.${process.env.NODE_ENV}`,
});

const Importer = require("../import");
const Promise = require("bluebird");
// const contentful = require('contentful-management')

describe("Incorrectly configured", () => {
  let TestImporter;
  beforeAll(async () => {
    TestImporter = new Importer(null, null, null);
  });
  afterEach(() => {});

  test("getSpace: rejects if invalid params used in configuration", async () => {
    // arrange

    // act

    // assert
    await expect(TestImporter.getSpace()).rejects.toThrow();
  });
  test("getEnvironment: rejects if invalid params used in configuration", async () => {
    // arrange
    // act
    // assert
    await expect(TestImporter.getEnvironment()).rejects.toThrow();
  });
  test("getContentTypes: throws error if space.getContentTypes rejects", async () => {
    // act
    // assert
    await expect(TestImporter.getContentTypes()).rejects.toThrow();
  });
});

describe("Correctly configured", () => {
  let TestImporter;

  beforeAll(() => {
    TestImporter = new Importer(
      process.env.ACCESS_TOKEN,
      process.env.SPACE_ID,
      process.env.SYS_ID
    );

    return TestImporter.init();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("init: space is initialised", async () => {
    // arrange
    let result;
    let expected = "logistics";
    // act
    result = TestImporter;
    // assert
    await expect(result.space).toBeDefined();
  });

  test("init: environment is initialised", async () => {
    // arrange
    let result = true;
    let expected = "master";
    // act
    result = TestImporter;
    // assert
    await expect(result.environment).toBeDefined();
  });

  test("importContentTypes: throws error if wrong args passed", async () => {
    // act
    // assert
    await expect(TestImporter.importContentTypes()).rejects.toThrow();
  });

  test("importContentTypes: resolves", async () => {
    // arrange
    let spy = jest.spyOn(TestImporter, "createContentType");
    let expected = "Done";
    let result;
    // console.log('environment', TestImporter.environment)

    // act
    result = await TestImporter.importContentTypes("../contenttypes.json");
    // assert
    expect(result[0]).toBe(expected);
    expect(spy).toHaveBeenCalledTimes(1);

    // spy.mockClear()
  });

  test("getContentTypes: response matches snapshot", async () => {
    // act
    // assert
    await expect(TestImporter.getContentTypes()).toMatchSnapshot();
  });

  test("readFile: rejects if incorrect argument used", async () => {
    // act
    // assert
    await expect(TestImporter.readFile()).rejects.toThrow();
  });

  test("readFile: rejects if non existant file used", async () => {
    // act
    // assert
    await expect(TestImporter.readFile("unknownFile.txt")).rejects.toThrow();
  });

  test("writeFile: calls writeFile mock and returns saved on success", async () => {
    // arrange
    let data = Buffer.from("Test data");
    let expected = "Saved";
    let testfile = "somefile.txt";
    let result = await TestImporter.writeFile(testfile, data);
    // act
    // assert
    expect(result.toString()).toBe(expected);
  });

  test("writeFile: rejects if incorrect argument used", async () => {
    // act
    // assert
    await expect(TestImporter.writeFile()).rejects.toThrow();
  });

  test("parseXml: throws error if something goes wrong", async () => {
    // arrange
    // act
    // assert
    await expect(TestImporter.parseXml()).rejects.toThrow();
  });

  test("parseXml: throws error if something goes wrong", async () => {
    // arrange
    let expected = { node: "content" };
    let result;
    // act
    result = await TestImporter.parseXml("<node>content</node>");
    // assert
    expect(result).toEqual(expected);
  });

  test("parseXml: throws error if you pass non xml string", async () => {
    // act
    await expect(TestImporter.parseXml(9)).rejects.toThrow();
  });

  test("createContentType: rejects if invalid input passed as args", async () => {
    // arrange
    let item = null;
    // act
    // assert
    await expect(TestImporter.createContentType(item)).rejects.toThrow();
  });

  test("createContentType: Should be called once and then done", async () => {
    // arrange
    let spy = jest.spyOn(TestImporter, "createContentType");
    let result;
    let expected = "Done";
    let types = [{ id: "test", name: "Test" }];
    const getResult = async () => {
      return Promise.all(
        types.map(item => TestImporter.createContentType(item))
      );
    };
    // act
    result = await getResult();
    // assert
    expect(spy.mock.calls[0][0]).toEqual(types[0]);
    expect(spy).toHaveBeenCalledTimes(1);
    expect(result.length).toEqual(1);
    expect(result[0]).toEqual(expected);

    // clear mock
    spy.mockRestore();
  });

  test("createContentType: Should be called three times, then Done", async () => {
    // arrange
    let spy = jest.spyOn(TestImporter, "createContentType");
    let result;
    let expected = "Done";
    let types = [
      { id: "1", name: "One" },
      { id: "2", name: "Two" },
      { id: "3", name: "Three" },
    ];
    const getResult = async () => {
      return Promise.all(
        types.map(item => TestImporter.createContentType(item))
      );
    };
    // act
    result = await getResult();
    // assert
    expect(spy.mock.calls[0][0]).toEqual(types[0]);
    expect(spy).toHaveBeenCalledTimes(3);
    expect(result.length).toEqual(3);
    expect(result[0]).toEqual(expected);
    // clear mock
    spy.mockRestore();
  });

  test("importWordpressData: should reject with incorrect args", async () => {
    let spy = jest.spyOn(TestImporter, "importWordpressData");

    await expect(TestImporter.importWordpressData()).rejects.toThrow();
    expect(spy).toHaveBeenCalledTimes(1);
  });
  test.skip("importWordpressData: should resolve data", async () => {
    let spy = jest.spyOn(TestImporter, "importWordpressData");
    let expected = { root: "content" };
    let result;

    result = await TestImporter.importWordpressData("testfile.xml");

    expect(result).toEqual(expected);
    expect(spy).toHaveBeenCalledTimes(1);
    // clear mock
    spy.mockRestore();
  });

  test("mapPost: should create an object with the en locale", () => {
    const data = require("../__mocks__/mockData.json");
    let result;
    let pages;

    // inject data, bypassing other routes
    TestImporter.data = data;
    // get only pages
    pages = TestImporter.getItemByType("page");

    result = TestImporter.mapPost(pages[0]);

    expect(result).toMatchSnapshot();
  });
  test("mapPost: should create an object with the nl locale", () => {
    const data = require("../__mocks__/mockData.json");
    let result;
    let pages;

    // inject data, bypassing other routes
    TestImporter.data = data;
    // get only pages
    pages = TestImporter.getItemByType("page");

    result = TestImporter.mapPost(pages[1]);
    expect(result).toMatchSnapshot();
  });
  test("mapPost: should handle all data without error", () => {
    const data = require("../__mocks__/mockData.json");
    let results;
    let pages;

    // inject data, bypassing other routes
    TestImporter.data = data;
    // get only pages
    pages = TestImporter.getItemByType("page");

    results = pages.map(page => {
      TestImporter.mapPost(page);
    });
    expect(results.length).toEqual(pages.length);
  });

  test.skip("createContentEntryWithId: should create entry for each page", async () => {
    const data = require("../__mocks__/mockData.json");
    let spy = jest.spyOn(TestImporter, "createContentEntryWithId");
    let expected = "Done";
    let result;
    let pages;

    TestImporter.data = data;
    pages = TestImporter.getItemByType("page");

    // console.log(data.rss.channel[0]['item'].length)
    // console.log(pages.length)

    result = await TestImporter.createContentEntryWithId("page", pages[0]);

    console.log("result ", result);
    expect(result.length).toEqual(pages.length);
    // clear mock
    spy.mockRestore();
  });
  test("createContentEntryWithId: should resolve after seconds", async () => {
    // arrange
    const data = require("../__mocks__/mockData.json");
    let spy = jest.spyOn(TestImporter, "createContentEntryWithId");
    let expected = "Done";
    let results;
    let pages;
    // jest.useFakeTimers()

    // act
    TestImporter.data = data;
    pages = TestImporter.getItemByType("page").slice(0, 5);

    // console.log(data.rss.channel[0]['item'].length)
    // console.log(pages.length)

    results = await Promise.map(
      pages,
      async page => {
        return TestImporter.createContentEntryWithId("page", page);
      },
      { concurrency: 1 }
    );

    // assert
    // expect(setTimeout).toHaveBeenCalledTimes(1)
    // expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000)
    expect(results.length).toEqual(pages.length);
    // clear mock
    spy.mockRestore();
  });
  test("getByNode, for author returns author element", () => {
    // arrange
    let results;
    let expected = 3;
    let data = require("../__mocks__/mockData.json");

    //  act
    // inject data, bypassing other routes
    TestImporter.data = data;
    results = TestImporter.getByNode("wp:author");
    // assert
    expect(results.length).toEqual(expected);
  });
  test("getByNode, returns an empty array if nothing found", () => {
    // arrange
    let result;
    let expected = [];
    let data = require("../__mocks__/mockData.json");

    //  act
    // inject data, bypassing other routes
    TestImporter.data = data;
    result = TestImporter.getByNode();
    // assert
    expect(result).toEqual(expected);
  });
  test("mapAuthor mapping should match snapshot", () => {
    // arrange
    let result;
    let data = require("../__mocks__/mockData.json");

    //  act
    // inject data, bypassing other routes
    TestImporter.data = data;
    result = TestImporter.getByNode("wp:author");
    result = TestImporter.mapAuthor(result[1]);
    // assert
    expect(result).toMatchSnapshot();
  });
});
