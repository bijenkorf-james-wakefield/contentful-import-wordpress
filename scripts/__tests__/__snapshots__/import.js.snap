// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly configured getContentTypes: response matches snapshot 1`] = `Promise {}`;

exports[`Correctly configured mapAuthor mapping should match snapshot 1`] = `
Object {
  "fields": Object {
    "display_name": Object {
      "en": "Queeny van Raaij",
    },
    "email": Object {
      "en": "Queeny.Raaijvan@deBijenkorf.nl",
    },
    "first_name": Object {
      "en": "Queeny",
    },
    "id": "2",
    "last_name": Object {
      "en": "van Raaij",
    },
    "slug": Object {
      "en": "queeny-van-raaij",
    },
  },
}
`;

exports[`Correctly configured mapPost: should create an object with the en locale 1`] = `
Object {
  "fields": Object {
    "body": Object {
      "en": Object {
        "content": Array [
          Object {
            "content": Array [
              Object {
                "data": Object {},
                "marks": Array [],
                "nodeType": "text",
                "value": "
				<ol class=\\"anchor-list\\">
<li><a href=\\"#content\\">Contents</a></li>
<li><a href=\\"#copyright\\">Copyright and trademarks</a></li>
<li><a href=\\"#personal\\">Personal information</a></li>
</ol>
<h2>Algemeen</h2>
Op de toegang tot en het gebruik van deze website van Magazijn de Bijenkorf B.V (de Bijenkorf) zijn, naast de toepasselijke wetgeving, de volgende bepalingen en voorwaarden van toepassing.
<ol>
<li id=\\"content\\">Contents
<ol>
<li>This web site has been prepared solely for the purpose of providing information about the logistic<a href=\\"http://www.clubink.se\\">clubink/a> processes of de Bijenkorf.</li>
<li>The information contained in this web site has been compiled in good faith by de Bijenkorf and is provided as is.</li>
<li>De Bijenkorf gives no representations or warranties, either express or implied, as to the <a href=\\"http://www.clubink.se\\">clubink.se</a> completeness, timeliness or accuracy of any information it contains.</li>
<li>Neither de Bijenkorf nor any of its officers, employees or agents shall be liable for any direct, incidental, consequential, indirect or punitive damages or losses arising out of any access to or use of or reliance upon (information contained in) this web site or any site linked to it, including but not limited to viruses that may infect computer equipment and/or software..</li>
<li>De Bijenkorf is not responsible for the contents of any off-site pages or any other web sites linked to or from this Bijenkorf Logistics web site.</li>
<li>De Bijenkorf may at any time and without notice, make any changes and corrections to the web site and to these terms and conditions.</li>
</ol>
</li>
<li id=\\"copyright\\">Copyright and trademarks
<ol>
<li>All copyright and other intellectual property rights in the information contained in this web site (including but not limited to text and images) are owned by de Bijenkorf.</li>
<li>Persons accessing this web site are permitted to display, store, reproduce and distribute extracts of this web site, but in all cases only for personal, informational and non-commercial purposes. No reproduction may be sold or distributed for commercial gain nor be modified or incorporated in any other work, publication or web site. No other rights or licenses are assigned or shall otherwise pass to persons accessing this web site.</li>
<li>All trademarks and emblems displayed on this web site are owned by de Bijenkorf.</li>
</ol>
</li>
<li id=\\"personal\\">Personal information
<ol>
<li>Visits to the site may be tracked through the use of 'cookies'. Cookies are pieces of information that a web site transfers to a computer's hard drive, for instance for record-keeping purposes. The use of cookies is an industry standard.</li>
<li>The information collected during use of this web site is used by de Bijenkorf to monitor the use of the web site and to help its further development.</li>
<li>De Bijenkorf never uses cookies to retrieve information from a computer that was not originally sent in a cookie.</li>
<li>Most browsers are initially set to accept cookies. If preferred, the browser settings can be changed to refuse cookies. However, in doing so, one may not be able to fully utilise the site.</li>
</ol>
</li>
</ol>
Access to and use of de Bijenkorf Logistics web site updated: June 28, 2016, subject to Internet legal notices.",
              },
            ],
            "data": Object {},
            "nodeType": "paragraph",
          },
        ],
        "data": Object {},
        "nodeType": "document",
      },
    },
    "id": "39",
    "slug": Object {
      "en": "disclaimer",
    },
    "title": Object {
      "en": "Disclaimer",
    },
  },
}
`;

exports[`Correctly configured mapPost: should create an object with the nl locale 1`] = `
Object {
  "fields": Object {
    "body": Object {
      "nl": Object {
        "content": Array [
          Object {
            "content": Array [
              Object {
                "data": Object {},
                "marks": Array [],
                "nodeType": "text",
                "value": "
				<h2>Distributiecentrum de Bijenkorf Woerden</h2>
Middellandbaan 1b
3447 GP Woerden
tel.: 088-245 4900
e-mail: <a href=\\"mailto:dbwmiddelland@debijenkorf.nl\\">deliveries@debijenkorf.nl</a>

&nbsp;
<h2>Magazijn de Bijenkorf B.V. Servicekantoor Amsterdam</h2>
Hoogoorddreef 11
1101 BA Amsterdam ZO

Postbus 12870
1100 AW Amsterdam ZO

tel.: 088-245 4000

BTW nummer: NL 003177397B01
EORI nummer: NL 003177397
Douanekantoor Nijmegen/Kantoor De Lutte - NL000851

<span style=\\"font-size: 9pt; color: #333333; font-family: Arial;\\">ABN AMRO BANK: </span><span style=\\"font-size: 9pt; font-family: Arial;\\"><span style=\\"color: #000000;\\">50.11.34.476</span></span><span style=\\"font-size: 8pt; color: red; font-family: Arial;\\"></span><span style=\\"font-size: 9pt; color: #333333; font-family: Arial;\\">
IBAN: </span><span style=\\"font-size: 10pt;\\"><span style=\\"color: #000000;\\">NL16ABNA0501134476</span></span><span style=\\"font-size: 8pt; color: red; font-family: Arial;\\"></span><span style=\\"font-size: 9pt; color: #333333; font-family: Arial;\\">
BIC: </span><span style=\\"font-size: 10pt;\\"><span style=\\"color: #000000;\\">ABNANL2A</span></span><span style=\\"font-size: 8pt; color: red; font-family: Arial;\\"></span>

<span style=\\"font-size: 9pt; color: #333333; font-family: Arial;\\">KvK nummer: 33116577</span>

&nbsp;
<h2>Monsterzendingen:</h2>
De Bijenkorf
Inkoop afdeling/naam inkoper
Hoogoorddreef 11
1101 BA Amsterdam ZO
tel.: 088-245 4000

&nbsp;",
              },
            ],
            "data": Object {},
            "nodeType": "paragraph",
          },
        ],
        "data": Object {},
        "nodeType": "document",
      },
    },
    "id": "282",
    "slug": Object {
      "nl": "contact",
    },
    "title": Object {
      "nl": "Contact",
    },
  },
}
`;
